types/eslint/index.d.ts

```diff
@@ -430,9 +430,16 @@ export namespace SourceCode {
 //#endregion
 
 export namespace Rule {
+    /**
+     * TODO: Old style rules are planned to be removed in v9, remove this type then (https://github.com/eslint/rfcs/blob/main/designs/2021-schema-object-rules/README.md)
+     * @deprecated Use `RuleModule` instead.
+     */
+    type OldStyleRule = Omit<RuleModule, "create"> & { create?: never } & RuleModule["create"];
```

Comment:
I believe the way the codebase works is that it actually first normalises the rule to be `{create: Function}` via the `getRuleFromConfig`, and then it passes this normalised object around.

I don't believe that it will actually treat a function-like-an-object and attempt to read things off of the function.

EG this type should just be `type OldStyleRule = RuleModule["create"]`